### PR TITLE
Show icons for various different file types

### LINKED ISSUE #226: new icons
Here's how the new icons would look:

Navigator, with inline breadcrumbs

Things that are changed:

1. Darker color (as they can be now, as they're not filled but outline based).
2. Basic mimetypes; including just those that are available and make sense for admins.
3. Using icon SVGs from ForkAwesome, which is a fork from FontAwesome. (PatternFly's icons are also forked from FontAwesome.)
4. Emblem for symlinks.

The icons should exist in one place and be referenced instead of being inline SVG. (Using inline SVG is not great for performance at the level of a thousand items, which some system directories have.)

### LINKED ISSUE #480: Classifying files as filetypes, for sorting and icon classification
> But what about the filetype sorting? Do we replicate nautilus behavior of showing the biggest (or smallest) files first and then folders?

Sorting by type is:

1. Folders first (Type: Folder)
2. General type alphanumeric (Archive for zip, tgz, bz2, etc.; Document for PDF, ODT, etc.; Image for PNG, JPEG, etc.)
3. Extension sorting, A-Z
4. Base filename sorting, A-Z

That's what Nautilus does, and it makes sense.

Nautilus has the mapping of types @ https://gitlab.gnome.org/GNOME/nautilus/-/blob/main/src/nautilus-mime-actions.c#L100-259

<details>
<summary>Here's the same thing in a JSON object</summary>

```json
{
    "Files": [
        "application/octet-stream",
        "text/plain",
        "text/markdown"
    ],
    "Folders": [
        "inode/directory"
    ],
    "Documents": [
        "application/rtf",
        "application/msword",
        "application/vnd.sun.xml.writer",
        "application/vnd.sun.xml.writer.global",
        "application/vnd.sun.xml.writer.template",
        "application/vnd.oasis.opendocument.text",
        "application/vnd.oasis.opendocument.text-template",
        "application/x-abiword",
        "application/x-applix-word",
        "application/x-mswrite",
        "application/docbook+xml",
        "application/x-kword",
        "application/x-kword-crypt",
        "application/x-lyx",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "text/markdown"
    ],
    "Illustration": [
        "application/illustrator",
        "application/vnd.corel-draw",
        "application/vnd.stardivision.draw",
        "application/vnd.oasis.opendocument.graphics",
        "application/x-dia-diagram",
        "application/x-karbon",
        "application/x-killustrator",
        "application/x-kivio",
        "application/x-kontour",
        "application/x-wpg"
    ],
    "Music": [
        "application/ogg",
        "audio/x-vorbis+ogg",
        "audio/ac3",
        "audio/basic",
        "audio/midi",
        "audio/x-flac",
        "audio/mp4",
        "audio/mpeg",
        "audio/x-mpeg",
        "audio/x-ms-asx",
        "audio/x-pn-realaudio"
    ],
    "PDF / PostScript": [
        "application/pdf",
        "application/postscript",
        "application/x-dvi",
        "image/x-eps",
        "image/vnd.djvu+multipage"
    ],
    "Picture": [
        "application/vnd.oasis.opendocument.image",
        "application/x-krita",
        "image/bmp",
        "image/cgm",
        "image/gif",
        "image/jpeg",
        "image/jpeg2000",
        "image/png",
        "image/svg+xml",
        "image/tiff",
        "image/x-compressed-xcf",
        "image/x-pcx",
        "image/x-photo-cd",
        "image/x-psd",
        "image/x-tga",
        "image/x-xcf"
    ],
    "Presentation": [
        "application/vnd.ms-powerpoint",
        "application/vnd.sun.xml.impress",
        "application/vnd.oasis.opendocument.presentation",
        "application/x-magicpoint",
        "application/x-kpresenter",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation"
    ],
    "Spreadsheet": [
        "application/vnd.lotus-1-2-3",
        "application/vnd.ms-excel",
        "application/vnd.stardivision.calc",
        "application/vnd.sun.xml.calc",
        "application/vnd.oasis.opendocument.spreadsheet",
        "application/x-applix-spreadsheet",
        "application/x-gnumeric",
        "application/x-kspread",
        "application/x-kspread-crypt",
        "application/x-quattropro",
        "application/x-sc",
        "application/x-siag",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
    ],
    "Text File": [
        "text/plain",
        "text/markdown"
    ],
    "Video": [
        "video/mp4",
        "video/3gpp",
        "video/mpeg",
        "video/quicktime",
        "video/vivo",
        "video/x-avi",
        "video/x-mng",
        "video/x-ms-asf",
        "video/x-ms-wmv",
        "video/x-msvideo",
        "video/x-nsv",
        "video/x-real-video"
    ]
}
```
</details>

I think we can do better though. Many of the mimetypes make sense; we can probably abstract some by searching the beginning of the mimetype string; for example, these match

Picture: `^image/`
Video: `^video/`
Music: `^audio/`

The easiest is probably just matching the first part of the mimetype in a first pass and extend it later.

_Originally posted by @garrett in https://github.com/cockpit-project/cockpit-files/issues/479#issuecomment-2137260196_

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Create a new esbuild plugin to compile data from various sources, creating an efficient map of filename extensions to file type information.

The most important piece of the filetype information at present, is a "class" attribute which is used to add an extra class to the `<tr>` containing the file in the list or grid views.  In turn, we add a number of icons for the various classes to our stylesheet.

This gets us special icons for "archive", "audio", "source code", "image", "text" and "video" files.

The file type classification is done entirely on the basis of the filename extension.  More specifically: there is no content sniffing.

Closes #226
Closes #480

### CODE REVIEW

File: src/app.tsx [hunk 1]
```diff
@@ -61,6 +65,31 @@
...
+        return null;
+    }
+
+    const last_dot = name.lastIndexOf('.');
+    if (last_dot !== 0 && last_dot !== -1 && last_dot + 20 > namelen) {
+        const second_last_dot = name.lastIndexOf('.', last_dot - 1);
+        if (second_last_dot !== -1 && second_last_dot !== 0 && second_last_dot + 20 > namelen) {
+            const ext = name.substring(second_last_dot + 1);
+            if (ext in filetype_data.exts) {
+                return filetype_data.cats[filetype_data.exts[ext]];
```
Comment (cockpituous) [inline src/app.tsx:81]: These 4 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+export interface FileTypeData {
+    cats: Record<Category, CategoryMetadata>;
+    exts: Record<string, Category>;
+}
+
+/* Doesn't actually do anything, but lets xgettext find the strings */
+function _(v: string) {
+    return v;
```
Comment (cockpituous) [inline src/filetype-plugin.ts:29]: These 2 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+export function create_filetype_data() : FileTypeData {
+    const cats = {
+        [Category.FILE]: { name: _("Unknown type"), icon: "file-o" },
+        [Category.ARCHIVE]: { name: _("Archive file"), icon: "file-archive-o" },
+        [Category.AUDIO]: { name: _("Audio file"), icon: "file-audio-o" },
+        [Category.CODE]: { name: _("Source code file"), icon: "file-code-o" },
+        [Category.IMAGE]: { name: _("Image file"), icon: "file-image-o" },
+        [Category.TEXT]: { name: _("Text file"), icon: "file-text-o" },
+        [Category.VIDEO]: { name: _("Video file"), icon: "file-video-o" },
+    } as const;
```
Comment (cockpituous) [inline src/filetype-plugin.ts:41]: These 10 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    const exts: { [ext: string]: Category } = {};
```
Comment (cockpituous) [inline src/filetype-plugin.ts:43]: This added line is not executed by any test.

File: src/filetype-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    function add_category_extensions(category: Category, extensions: string[]) {
+        for (let ext of extensions) {
+            if (ext[0] === '.') {
+                ext = ext.substring(1);
```
Comment (cockpituous) [inline src/filetype-plugin.ts:48]: These 4 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            }
+            exts[ext] = category;
```
Comment (cockpituous) [inline src/filetype-plugin.ts:50]: This added line is not executed by any test.

File: src/filetype-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    for (const [mimetype, details] of Object.entries(mime_db)) {
+        if ('extensions' in details) {
+            if (mimetype.startsWith('image/')) {
+                add_category_extensions(Category.IMAGE, details.extensions);
+            } else if (mimetype.startsWith('audio/')) {
+                add_category_extensions(Category.AUDIO, details.extensions);
+            } else if (mimetype.startsWith('video/')) {
+                add_category_extensions(Category.VIDEO, details.extensions);
+            } else if (mimetype.startsWith('text/')) {
+                add_category_extensions(Category.TEXT, details.extensions);
```
Comment (cockpituous) [inline src/filetype-plugin.ts:64]: These 10 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        }
+    }
+
+    // Archives, scraped from https://en.wikipedia.org/wiki/List_of_archive_formats
+    add_category_extensions(
+        Category.ARCHIVE,
```
Comment (cockpituous) [inline src/filetype-plugin.ts:71]: These 2 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        [
+            '7z', 'F', 'LBR', 'Z', 'a', 'aar', 'ace', 'afa', 'alz', 'apk', 'ar', 'arc', 'arc', 'arj', 'ark', 'b1',
+            'b6z', 'ba', 'bh', 'br', 'bz2', 'cab', 'car', 'cdx', 'cfs', 'cpio', 'cpt', 'dar', 'dd', 'dgc', 'ear', 'gca',
+            'genozip', 'genozip', 'gz', 'ha', 'hki', 'ice', 'iso', 'jar', 'kgb', 'lbr', 'lha', 'lz', 'lz4', 'lzh',
+            'lzma', 'lzo', 'lzx', 'mar', 'pak', 'paq6', 'paq7', 'paq8', 'partimg', 'pea', 'phar', 'pim', 'pit', 'qda',
+            'rar', 'rk', 'rz', 's7z', 'sbx', 'sda', 'sea', 'sen', 'sfark', 'sfx', 'shar', 'shk', 'sit', 'sitx', 'sqx',
+            'sz', 'tar', 'tar.Z', 'tar.bz2', 'tar.gz', 'tar.lz', 'tar.xz', 'tar.zst', 'tbz2', 'tgz', 'tlz', 'txz', 'uc',
+            'uc0', 'uc2', 'uca', 'ucn', 'ue2', 'uha', 'ur2', 'war', 'wim', 'xar', 'xp3', 'xz', 'yz1', 'z', 'zip',
+            'zipx', 'zoo', 'zpaq', 'zst', 'zz'
```
Comment (cockpituous) [inline src/filetype-plugin.ts:80]: These 8 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        ]
+    );
+
+    // Special treatment for programming languages based on GitHub's database
+    for (const lang of Object.values(language_map)) {
+        if (lang.type === 'programming' && 'extensions' in lang) {
+            add_category_extensions(Category.CODE, lang.extensions);
```
Comment (cockpituous) [inline src/filetype-plugin.ts:87]: These 3 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    return { cats, exts };
```
Comment (cockpituous) [inline src/filetype-plugin.ts:91]: This added line is not executed by any test.

File: src/filetype-plugin.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+export function to_code(data: FileTypeData): string {
+    return `
```
Comment (cockpituous) [inline src/filetype-plugin.ts:95]: These 2 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        import cockpit from "cockpit";
+        const _ = cockpit.gettext;
+        export const filetype_data = {
+            cats: {
+                ${Object.entries(data.cats).map(([key, value]) =>
+        `${key}: {name: _(${JSON.stringify(value.name)}), icon: ${JSON.stringify(value.icon)}},`)
+            .join('\n')}
```
Comment (cockpituous) [inline src/filetype-plugin.ts:102]: These 3 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+            },
+            exts: ${JSON.stringify(data.exts)}
```
Comment (cockpituous) [inline src/filetype-plugin.ts:104]: This added line is not executed by any test.

File: src/filetype-plugin.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    return t.cats[t.exts[ext] || Category.FILE];
+}
+
+export class FileTypePlugin implements Plugin {
+    name = 'cockpit-files-filetype-plugin';
+
+    setup(build: PluginBuild) {
+        build.onResolve({ filter: /^\.\/filetype-data$/ }, args => ({
+            path: args.path,
+            namespace: 'cockpit-files-filetype-plugin',
```
Comment (cockpituous) [inline src/filetype-plugin.ts:119]: These 4 added lines are not executed by any test.

File: src/filetype-plugin.ts [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        }));
+
+        build.onLoad({ filter: /.*/, namespace: 'cockpit-files-filetype-plugin' }, () => ({
+            contents: to_code(create_filetype_data()),
+            resolveDir: 'pkg/lib',
```
Comment (cockpituous) [inline src/filetype-plugin.ts:124]: These 3 added lines are not executed by any test.

File: build.js [hunk 17]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env node
+#!/usr/bin/env -S node --import tsx/esm
```
Comment (jelly) [inline build.js:1]: What is this exactly?

File: src/app.tsx [hunk 18]
```diff
@@ -108,6 +137,7 @@ export const Application = () => {
                 const files = entries.map(([name, attrs]) => ({
                     ...attrs,
                     name,
+                    category: categorize(name, attrs),
                     to: info.target(name)?.type ?? null
```
Comment (jelly) [inline src/app.tsx:141]: Right, we wanna re-do this at some point, but I suppose that won't change much for the function definitiion

File: src/files-card-body.jsx [hunk 19]
```diff
@@ -422,9 +422,13 @@ const Item = React.memo(function Item({ file, isSelected, isGrid }) {
...
-                    {file.type === "dir" || file.to === "dir"
-                        ? <FolderIcon />
-                        : <FileIcon />}
+                    {((file.type === "dir" || file.to === "dir") && <FolderIcon />) ||
+                        (file.category &&
+                        <img
+                          width='54px' height='54px'
+                          src={'icons/' + file.category.icon + '.svg'}
+                          aria-label={file.category.name}
+                        />) || <FileIcon />}
```
Comment (jelly) [inline src/files-card-body.jsx:431]: This conflicts with the work I put in yesterday where the whole ListingTable is used for both views and the svg's 
are embedded in scss.

We'll need to investigate what / how to tackle that. Either is fine but we must make sure the svg's are cached by the browser just as JS/CSS.

https://github.com/cockpit-project/cockpit-files/pull/481

File: src/sidebar.jsx [hunk 20]
```diff
@@ -72,6 +72,16 @@ const getDescriptionListItems = selected => {
...
+                {
+                    id: "description-list-category",
+                    label: _("Category"),
+                    value: selected.category.name
+                },
+                {
+                    id: "description-list-icon",
+                    label: _("Icon"),
+                    value: selected.category.icon
+                },
```
Comment (jelly) [inline src/sidebar.jsx:84]: This is in no design, we just want icons in the list view and overview. Can we keep things simple please :)

File: src/filetype-plugin.ts [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    return { cats, exts };
+}
+
+export function to_code(data: FileTypeData): string {
+    return `
+        import cockpit from "cockpit";
+        const _ = cockpit.gettext;
+        export const filetype_data = {
```
Comment (jelly) [inline src/filetype-plugin.ts:98]: So this is a dynamic module which ends up in `dist`? That is way better then how our translations work by dynamically injecting them as `.js`.

I still feel this is rather overblown

File: src/filetype-plugin.ts [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+            cats: {
```
Comment (jelly) [inline src/filetype-plugin.ts:99]: `cats` doesn't autocomplete to categories in my head. We don't have to safe space here

File: src/filetype-plugin.ts (hunk 14 above)
Comment (jelly) [inline src/filetype-plugin.ts:104]: Why are we converting this to a string? It's a JS module right? So it can just use native types?

Comment (jelly) [reply src/filetype-plugin.ts:104]: I can answer that myself, we are building a JS file here...

**Review by jelly — COMMENTED** [review]
This feels a bit slightly overengineered for 4 icons :)

Please add a small description in the future to the draft PR. As far as I can see this is an ESBuild plugin which generates a file with "file type" information, shipped and used by client code to determine the icon for the given filetype.

**Comment by allisonkarlitskaya** [comment]
> This feels a bit slightly overengineered for 4 icons :)

Yup!  It was fun, though :)

File: src/filetype-plugin.ts (hunk 14 above)
Comment (allisonkarlitskaya) [reply src/filetype-plugin.ts:104]: There are two tricky parts here that doesn't let us handle this as straight JSON (which would have been my first choice):
 - our `Category.*` constants are integers and you can't have integer keys in JSON.  That's minor, and we could have used strings to workaround that.
 - less minor: we need the `_()` evaluated in the browser at run-time to make sure the category strings get translated.  If we end up never showing those in the browser then (and don't care about the sort order being locale-sensitive) then we could maybe ignore this point as well...

File: build.js (hunk 17 above)
Comment (allisonkarlitskaya) [reply build.js:1]: It lets us write esbuild plugins in Typescript by enabling `node` to `import` `.ts` files.

File: src/app.tsx (hunk 18 above)
Comment (allisonkarlitskaya) [reply src/app.tsx:141]: We're always going to need to do some kind of "special treatment" for symlinks here, so however we handle that, the category code will be part of the same.

File: src/files-card-body.jsx (hunk 19 above)
Comment (allisonkarlitskaya) [reply src/files-card-body.jsx:431]: I absolutely hate this part of the patch and definitely want you or @garrett to come up with something better here.

File: src/sidebar.jsx (hunk 20 above)
Comment (allisonkarlitskaya) [reply src/sidebar.jsx:84]: Ya.  This is purely debugging output.  When I said this is "proof of concept" I really meant it :)

File: src/filetype-plugin.ts (hunk 21 above)
Comment (allisonkarlitskaya) [reply src/filetype-plugin.ts:98]: It ends up in the bundle, ya.  One of the big things I'd like to find out is if we can avoid translating the "Image", "Audio", etc. part.  If @garrett only wants those for sorting and not showing to the user then probably the exact description of the string in the local language isn't important.

That would let us encode this thing as JSON, which would make the whole approach a fair bit less offensive in my opinion (ie: we could kill this `to_code()` crud).

**Comment by allisonkarlitskaya** [comment]
Also: feedback to `<object>`: if I use that instead of `` I can't get the icons to scale at all, plus the icon is no longer "clickable" to select the file.  With `` it works.

File: src/filetype-plugin.ts (hunk 14 above)
Comment (allisonkarlitskaya) [reply src/filetype-plugin.ts:104]: I changed this to JSON.  I feel like we can deal with the translation in a better way (if we even need to).  Also: the thing about the numbers is a non-issue.  Hilariously, this works:

```
Welcome to Node.js v20.12.2.
Type ".help" for more information.
> {"0":1}[0]
1
```

so the quoting that happens is completely irrelevant.

File: src/app.tsx [hunk 23]
```diff
@@ -99,17 +103,18 @@ export const Application = () => {
...
                 setCwdInfo(state.info);
                 setErrorMessage(state.error?.message ?? "");
                 const entries = Object.entries(state?.info?.entries || {});
-                const files = entries.map(([name, attrs]) => ({
-                    ...attrs,
-                    name,
-                    to: info.target(name)?.type ?? null
-                }));
+                const files = entries.map(([name, attrs]) => {
+                    const to = info.target(name)?.type ?? null;
```
Comment (cockpituous) [inline src/app.tsx:114]: This added line is not executed by any test.

File: src/files-card-body.jsx [hunk 24]
```diff
@@ -422,9 +422,13 @@ const Item = React.memo(function Item({ file, isSelected, isGrid }) {
...
-                    {file.type === "dir" || file.to === "dir"
-                        ? <FolderIcon />
-                        : <FileIcon />}
+                    {((file.type === "dir" || file.to === "dir") && <FolderIcon />) ||
+                        (file.category &&
+                        <img
+                          width='48px' height='48px'
+                          src={'icons/' + file.category.icon + '.svg'}
+                          aria-label={file.category.name}
+                        />) || <FileIcon />}
```
Comment (cockpituous) [inline src/files-card-body.jsx:431]: This added line is not executed by any test.

File: src/app.tsx (hunk 23 above)
Comment (cockpituous) [inline src/app.tsx:114]: This added line is not executed by any test.

File: src/files-card-body.jsx [hunk 25]
```diff
@@ -342,94 +351,97 @@ export const FilesCardBody = ({
         </ContextMenu>
     );
 
+    const sortColumn = (columnIndex) => ({
+        sortBy: {
+            index: filterColumnMapping[sortBy][0],
+            direction: filterColumnMapping[sortBy][1],
+        },
+        onSort: (_event, index, direction) => {
+            setSortBy(filterColumns[index][direction].itemId);
```
Comment (cockpituous) [inline src/files-card-body.jsx:360]: These 2 added lines are not executed by any test.

File: src/files-card-body.jsx [hunk 26]
```diff
@@ -342,94 +351,97 @@ export const FilesCardBody = ({
...
-        } else {
-            return "file-item";
-        }
+const getFileType = (file) => {
+    if (file.type === "dir" || file.to === "dir") {
+        return "folder";
+    } else if (file.category.class) {
+        return file.category.class;
+    } else {
+        return "file";
```
Comment (cockpituous) [inline src/files-card-body.jsx:408]: This added line is not executed by any test.

File: src/app.tsx (hunk 23 above)
Comment (cockpituous) [inline src/app.tsx:114]: This added line is not executed by any test.

File: build.js (hunk 17 above)
Comment (jelly) [reply build.js:1]: Grrmpfft why is that not the default.

File: src/filetype-lookup.ts [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+export interface FileTypeData {
+    categories: Record<Category, CategoryMetadata>;
+    extensions: Record<string, Category>;
+    max_extension_length: number;
+}
+
+export function filetype_lookup(filetype_data: FileTypeData, name: string) {
+    // Never find a dot at offset '0' (the one for a hidden file), or one that
+    // would produce an extension that we know not to be in the database.
```
Comment (jelly) [inline src/filetype-lookup.ts:25]: `..` and `.` would also not really be useful to check.

File: src/filetype-lookup.ts [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+    let offset = Math.max(1, name.length - filetype_data.max_extension_length - 1);
+
+    // Allow finding extensions like '.tar.gz' (prefer longest first)
+    while ((offset = name.indexOf('.', offset)) > 0) {
```
Comment (jelly) [inline src/filetype-lookup.ts:29]: In reality we only need to go two levels deep right? `.tar.gz`

File: src/filetype-plugin.ts [hunk 29]
```diff
@@ -0,0 +1,88 @@
+import { Plugin, PluginBuild } from 'esbuild';
```
Comment (jelly) [inline src/filetype-plugin.ts:1]: I kinda would like to have a small description about this magic. As there is no file created after this plugin has build which is something new and something we never really did before.

File: src/filetype-plugin.ts [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+
+    setup(build: PluginBuild) {
+        build.onResolve({ filter: /^\.\/filetype-data$/ }, args => ({
+            path: args.path,
+            namespace: 'cockpit-files-filetype-plugin',
+        }));
+
+        build.onLoad({ filter: /.*/, namespace: 'cockpit-files-filetype-plugin' }, () => ({
+            contents: JSON.stringify(create_filetype_data()),
+            loader: 'json',
```
Comment (jelly) [inline src/filetype-plugin.ts:83]: So the plugin just exports the created JSON, super interesting :)

File: src/app.tsx (hunk 23 above)
Comment (cockpituous) [inline src/app.tsx:114]: This added line is not executed by any test.

File: src/filetype-plugin.ts (hunk 29 above)
Comment (allisonkarlitskaya) [reply src/filetype-plugin.ts:27]: Added

File: src/filetype-plugin.ts (hunk 30 above)
Comment (allisonkarlitskaya) [reply src/filetype-plugin.ts:109]: Yup.  This esbuild plugin stuff is actually pretty straightforward...

**Pull request merged**